File: haskell/blog-server/src/Domain/UseCases/Works/Get/Input.hs
module Domain.UseCases.Works.Get.Input
  ( GetWorkInput (..),
  )
where

import Data.Text (Text)
import Domain.Entities.Work (Work)

type Slug = Text

data GetWorkInput = GetWorkInput
  { _slug :: Slug,
    _readWork :: Slug -> IO (Either Text Work)
  }

File: fsharp/src/WorksServer/UseCases/Works/Redeploy/Input.fs
module WorksServer.UseCases.Works.Redeploy.Input

open WorksServer.Entities.Work

type Page =
    { id: string
      slug: string
      category: string
      title: string
      description: string
      thumbnailUrl: string
      languages: string list
      publishedAt: string option
      unpublishedAt: string option
      createdAt: string
      updatedAt: string }

type FetchAllPagesFromCMS = Async<Page list>

type FetchPageBodyFromCMS = string -> Async<string>

type UpdateOrCreateWork = Work -> Async<Result<int, string>>

type RedeployWorksInput =
    { fetchAllPageFromCMS: FetchAllPagesFromCMS
      fetchPageBodyFromCMS: FetchPageBodyFromCMS
      updateOrCreateWork: UpdateOrCreateWork }
